e2e_tests/tests/cluster/test_ship_logs.py

```diff
@@ -0,0 +1,420 @@
+        )
+        assert isinstance(exit_code, int), exit_code
+        return exit_code
+
+    def mkcmd(self, script: str) -> List[str]:
+        # -u: don't buffer stdout/stderr.
+        return [sys.executable, "-u", "-c", textwrap.dedent(script)]
+
+    @pytest.mark.e2e_cpu
+    def test_preserve_exit(self) -> None:
+        cmd = self.mkcmd(
+            """
+            import sys
+            print("hi", file=sys.stdout, flush=True)
+            print("bye", file=sys.stderr, flush=True)
+            sys.exit(9)
+            """
+        )
+        with ShipLogServer() as srv:
+            exit_code = self.run_ship_logs(srv.master_url(), cmd)
+        assert exit_code == 9, exit_code
+        # Ordering of stdout vs stderr is non-deterministic.
+        assert set(srv.logs) == {"hi\n", "bye\n"}, srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_cr_to_lf(self) -> None:
+        cmd = self.mkcmd(
+            r"""
+            print("1\n", end="")
+            print("2\r", end="")
+            print("3\r\n", end="")
+            """
+        )
+        with ShipLogServer() as srv:
+            exit_code = self.run_ship_logs(srv.master_url(), cmd)
+        assert exit_code == 0, exit_code
+        assert "".join(srv.logs) == "1\n2\n3\n\n", srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_stdout_stderr_ordering(self) -> None:
+        # Stdout and stderr are collected on different threads, and therefore _can't_ be perfectly
+        # synced.  But they should be "approximately" synced; i.e. each 1-second batch should
+        # contain both log types.
+        #
+        # Most dev machines probably will be fine with small timeouts, but CI machines might be
+        # slower and we allow up to 0.2 seconds of slop.
+        timeouts = [0.001, 0.2]
+        for timeout in timeouts:
+            cmd = self.mkcmd(
+                f"""
+                import sys
+                import time
+                print("1", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("2", file=sys.stderr, flush=True)
+                time.sleep({timeout})
+                print("3", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("4", file=sys.stderr, flush=True)
+                time.sleep({timeout})
+                print("5", file=sys.stdout, flush=True)
+                time.sleep({timeout})
+                print("6", file=sys.stderr, flush=True)
+                """
+            )
+            with ShipLogServer() as srv:
+                exit_code = self.run_ship_logs(srv.master_url(), cmd)
+            assert exit_code == 0, exit_code
+            if "".join(srv.logs) == "1\n2\n3\n4\n5\n6\n":
+                # Success
+                break
+            elif timeout == timeouts[-1]:
+                # Failed, even on the highest timeout
+                raise ValueError("".join(srv.logs))
+
+    @pytest.mark.e2e_cpu
+    def test_signal_forwarding(self) -> None:
```

Comment:
This is going to be a little meandering...

I wrote [an opinionated dev guide](https://hpe-aiatscale.atlassian.net/wiki/spaces/ENGINEERIN/pages/1351909582/Python+Testing) for unit tests that advocates for really long, specific names.

>Tests should generally be named test_<method_under_test>_<behavior> (ex: test_get_versions_reads_paginated_response

These seem like a mix of unit tests and end-to-end (end to end from the log-shippers perspective, but isolated from the system they live on), and in that don't apply so clearly to "method under test".

I guess the immediacy of "method" and "test" is one of the reasons I especially like unit tests. But I also get that there are reasons to prefer end-to-end tests.

So, yeah... meandering. I think I'd like teams call or something to talk about what to isolate and what to test.

These tests do a good job of testing one particular thing. I want people to copy this approach when they write their own tests. And I think, all other things being equal, that I'd like the name to be more verbose for what behavior, exactly, they're testing. When tests are specific, it helps their writer themselves get specific about what's being tested. I think if that were also a part of these tests that'd help people get the "best-practices" content I hope they get when they use this as a guide.

Reply:
yeah, let's do a teams call.

Reply:
We talked yesterday, and we resolved some names on `test_killed_process` and `test_process_closed_stdout`.

But this comment is about `test_signal_forwarding`?  Do we still need to discuss this test specifically?

Reply:
This is just style, but I think names like `test_signals_are_forwarded` are slightly clearer to an oncall about what's broken if/when something breaks.

The recommendation is also here because as style, it makes a tester think harder about what's actually being tested and both keeps tests focused to a behavior while also ensuring that tests actually test something the author cares about.

This is a good test. It doesn't need a better name to make it a good test. I prefer the predicate-focused style here as part of an effort to make that style the norm so that bad tests become less common.

Reply:
done